core/src/main/scala/com/rallyhealth/vapors/factfilter/data/Evidence.scala

```diff
@@ -0,0 +1,225 @@
+import cats.{Eq, Monoid}
+import com.rallyhealth.vapors.core.logic.{Intersect, Union}
+
+/**
+  * The result of evaluating an [[com.rallyhealth.vapors.factfilter.dsl.Exp]]
+  */
+sealed abstract class Evidence extends Equals {
+
+  /**
+    * A subset of the given facts that can be used to prove the logical query to be true.
+    */
+  def facts: List[Fact]
+
+  /**
+    * True when [[facts]] is empty, and there is no subset of the facts provided that
+    * can sufficiently prove the query to be true.
+    */
+  final def isEmpty: Boolean = facts.isEmpty
+
+  /**
+    * True when [[facts]] is not empty, and there exists a subset of the provided facts
+    * that can be used to prove the query to be true.
+    */
+  @inline final def nonEmpty: Boolean = !isEmpty
+
+  /**
+    * Same as [[facts]], but as a Set.
+    */
+  lazy val factSet: Set[Fact] = facts.toSet
+
+  /**
+    * Union of the two sets of results.
+    */
+  final def union(o: Evidence): Evidence = {
+    if (o.factSet == this.factSet) o
+    else
+      this match {
+        case NoEvidence() => o
+        case FoundEvidence(matchingFactsNel) =>
+          val newFacts = o.facts.collect {
+            case fact if !factSet.contains(fact) => fact
+          }
+          FoundEvidence(matchingFactsNel ++ newFacts)
+      }
+  }
+
+  /**
+    * Alias for [[union]].
+    */
+  @inline final def ++(o: Evidence): Evidence = this.union(o)
+
+  /**
+    * Alias for [[union]].
+    */
+  @inline final def |(o: Evidence): Evidence = this.union(o)
+
+  // SAFE: Evidence is sealed, so all subclasses are matched on by this parent type.
+  def canEqual(other: Any): Boolean = other.isInstanceOf[Evidence]
+
+  override def equals(other: Any): Boolean = other match {
+    case that: Evidence =>
+      that.canEqual(this) &&
+        this.factSet == that.factSet
+    case _ => false
+  }
+
+  override def hashCode(): Int = factSet.hashCode()
+
+  override final def toString: String = this match {
+    case NoEvidence() => "NoEvidence()"
+    case matched: FoundEvidence => s"FoundEvidence(${matched.factsNel})"
+  }
+}
+
+object Evidence {
+
+  def none: Evidence = NoEvidence
+
+  def apply(facts: List[Fact]): Evidence = {
+    NonEmptyList.fromList(facts).map(FoundEvidence(_)).getOrElse(NoEvidence())
+  }
+
+  def fromNel(nonEmptyFacts: NonEmptyList[Fact]): FoundEvidence = FoundEvidence(nonEmptyFacts)
+
+  // TODO: This is picked up on every foldMap... is it reasonable to assume that all evidence of operations
+  //       like addition and subtraction always union their corresponding evidence?
+  //       Maybe it should reduce the evidence to zero when any value has zero evidence?
+  //       But then how would that work with adding constants?
+  implicit val monoid: Monoid[Evidence] = {
+    new Monoid[Evidence] {
+      override def empty: Evidence = NoEvidence
+      override def combine(
+        x: Evidence,
+        y: Evidence,
+      ): Evidence = x ++ y
+    }
+  }
```

Comment:
This is probably one of the biggest questions I need to answer about this whole evidence tracking... how should I handle operations like addition, subtraction, etc. It's an interesting academic exercise to figure this out for `And`, `Or`, `Not`, etc. But what about `ExistsInOutput`, `ForAllInOutput`, `FlatMapOutput`, etc?

Reply:
I'm going to investigate supporting evidence tracking for constants. For now, the docs on this are much clearer.